Fatras/include/ActsFatras/EventData/Barcode.hpp

```diff
@@ -91,81 +94,166 @@ namespace ActsFatras {
+  using ValueVtx = std::uint16_t;
+  using ValuePart = std::uint32_t;
+  using ValueGen = std::uint8_t;
 
   // Construct an invalid barcode with all levels set to zero.
-  constexpr Barcode() : Base(Base::Zeros()) {}
+  constexpr Barcode()
+      : vertexPrimaryID(0u),
+        vertexSecondaryID(0u),
+        particleID(0u),
+        generationID(0u),
+        subParticleID(0u) {}
+  // Construct a valid barcode.
+  constexpr Barcode(ValueVtx pv, ValueVtx sv, ValuePart p, ValueGen g,
+                    ValuePart sp = 0u)
+      : vertexPrimaryID(pv),
+        vertexSecondaryID(sv),
+        particleID(p),
+        generationID(g),
+        subParticleID(sp) {}
+  Barcode(const std::vector<std::uint32_t>& data) { setData(data); }
+  // Construct a valid barcode for vertex.
+  constexpr Barcode(ValueVtx pv, ValueVtx sv, ValueGen g)
+      : vertexPrimaryID(pv),
+        vertexSecondaryID(sv),
+        particleID(0u),
+        generationID(g),
+        subParticleID(0u) {}
+
   Barcode(const Barcode&) = default;
   Barcode(Barcode&&) = default;
   Barcode& operator=(const Barcode&) = default;
   Barcode& operator=(Barcode&&) = default;
 
+  friend constexpr bool operator<(Barcode lhs, Barcode rhs) {
+    if (lhs.vertexPrimary() != rhs.vertexPrimary()) {
+      return lhs.vertexPrimary() < rhs.vertexPrimary();
+    }
+    if (lhs.vertexSecondary() != rhs.vertexSecondary()) {
+      return lhs.vertexSecondary() < rhs.vertexSecondary();
+    }
+    if (lhs.particle() != rhs.particle()) {
+      return lhs.particle() < rhs.particle();
+    }
+    if (lhs.generation() != rhs.generation()) {
+      return lhs.generation() < rhs.generation();
+    }
+    return lhs.subParticle() < rhs.subParticle();
+  }
+
+  friend constexpr bool operator==(Barcode lhs, Barcode rhs) {
+    return lhs.vertexPrimary() == rhs.vertexPrimary() &&
+           lhs.vertexSecondary() == rhs.vertexSecondary() &&
+           lhs.particle() == rhs.particle() &&
+           lhs.generation() == rhs.generation() &&
+           lhs.subParticle() == rhs.subParticle();
+  }
+
   /// Return the primary vertex identifier.
-  constexpr Value vertexPrimary() const { return level(0); }
+  constexpr ValueVtx vertexPrimary() const { return vertexPrimaryID; }
   /// Return the secondary vertex identifier.
-  constexpr Value vertexSecondary() const { return level(1); }
+  constexpr ValueVtx vertexSecondary() const { return vertexSecondaryID; }
   /// Return the particle identifier.
-  constexpr Value particle() const { return level(2); }
+  constexpr ValuePart particle() const { return particleID; }
   /// Return the generation identifier.
-  constexpr Value generation() const { return level(3); }
+  constexpr ValueGen generation() const { return generationID; }
   /// Return the sub-particle identifier.
-  constexpr Value subParticle() const { return level(4); }
+  constexpr ValuePart subParticle() const { return subParticleID; }
+
+  std::vector<std::uint32_t> getData() const {
```

Comment:
```suggestion
  std::array<std::uint32_t, 5> asVector() const {
```
typedef might be good e.g.
```c++
using IdVector = std::array<std::uint32_t, 5>
```